parts/k8s/cloud-init/artifacts/cse_config.sh

```diff
@@ -387,14 +387,30 @@ ensureKubelet() {
     sed -i "s|<advertiseAddr>|$PRIVATE_IP|g" $f
   fi
   wait_for_file 1200 1 /opt/azure/containers/kubelet.sh || exit {{GetCSEErrorCode "ERR_FILE_WATCH_TIMEOUT"}}
-  {{- if HasKubeReservedCgroup}}
+{{- if HasKubeReservedCgroup}}
   wait_for_file 1200 1 /etc/systemd/system/{{- GetKubeReservedCgroup -}}.slice || exit {{GetCSEErrorCode "ERR_KUBERESERVED_SLICE_SETUP_FAIL"}}
   wait_for_file 1200 1 /etc/systemd/system/kubelet.service.d/kubereserved-slice.conf || exit {{GetCSEErrorCode "ERR_KUBELET_SLICE_SETUP_FAIL"}}
-  {{- end}}
-  systemctlEnableAndStart kubelet || exit {{GetCSEErrorCode "ERR_KUBELET_START_FAIL"}}
+{{- end}}
+  if [[ -n ${MASTER_NODE} ]]; then
+    systemctlEnableAndStart kubelet || exit {{GetCSEErrorCode "ERR_KUBELET_START_FAIL"}}
+  else
+{{- if not RunUnattendedUpgrades}}
+    systemctlEnableAndStart kubelet || exit {{GetCSEErrorCode "ERR_KUBELET_START_FAIL"}}
+{{else}}
+    retrycmd 120 5 25 systemctl enable kubelet || exit {{GetCSEErrorCode "ERR_KUBELET_START_FAIL"}}
+{{- end}}
+  fi
 {{- if HasKubeletHealthZPort}}
   wait_for_file 1200 1 /etc/systemd/system/kubelet-monitor.service || exit {{GetCSEErrorCode "ERR_FILE_WATCH_TIMEOUT"}}
-  systemctlEnableAndStart kubelet-monitor || exit {{GetCSEErrorCode "ERR_KUBELET_START_FAIL"}}
+  if [[ -n ${MASTER_NODE} ]]; then
+    systemctlEnableAndStart kubelet-monitor || exit {{GetCSEErrorCode "ERR_KUBELET_START_FAIL"}}
+  else
+  {{- if not RunUnattendedUpgrades}}
+    systemctlEnableAndStart kubelet-monitor || exit {{GetCSEErrorCode "ERR_KUBELET_START_FAIL"}}
+  {{else}}
+    retrycmd 120 5 25 systemctl enable kubelet-monitor || exit {{GetCSEErrorCode "ERR_KUBELET_START_FAIL"}}
```

Comment:
You likely need a systemctl daemon-reload here before the enable.  Just above you have just added the service file (potentially) and thus will need that reload to make systemctl know about it.  Your function for systemctlEnableAndStart calls systemctl_restart  first which does the systemctl daemon-reload.

(This is based on the fact that line 404 waits for the file to show up)

Reply:
Good catch, thanks!